kpatch-build/lookup.c

```diff
@@ -48,12 +48,77 @@ struct lookup_table {
 	int fd, nr;
 	Elf *elf;
 	struct symbol *syms;
+	struct symbol *local_syms;
 };
 
 #define for_each_symbol(ndx, iter, table) \
 	for (ndx = 0, iter = table->syms; ndx < table->nr; ndx++, iter++)
 
-struct lookup_table *lookup_open(char *path)
+static void resolve_dup_file_symbol(struct lookup_table *table, char *hint,
+				    struct sym_compare_type *locals)
+{
+	struct symbol *sym, *prev;
+	int i, in_file = 0;
+	int multi_base = 0;
+	struct sym_compare_type *local_index;
+
+	for_each_symbol(i, sym, table) {
+		if (sym->type == STT_FILE) {
+			if (in_file) {
+				if (table->local_syms) {
+					multi_base = 1;
+					in_file = 0;
+					break;
+				}
+				table->local_syms = prev;
+			}
+
+			if (!strcmp(hint, sym->name)) {
+				in_file = 1;
+				prev = sym;
+				local_index = locals;
+			}
+			else
+				in_file = 0;
+
+			continue;
+		}
+
+		if (!in_file)
+			continue;
+		if (sym->bind != STB_LOCAL || (sym->type != STT_FUNC && sym->type != STT_OBJECT))
+			continue;
+
+		for ( ; local_index->name; local_index++)
+			if (local_index->type == sym->type &&
+			    !strcmp(local_index->name, sym->name))
+				break;
```

Comment:
Why is there a loop here?  Shouldn't the symbols match up one-to-one between locals and sym?  For example, wouldn't something like this work instead of the for loop?
```
        if (local_index->type == sym->type && !strcmp(local_index->name, sym->name))
                local_index++;
        else
                in_file = 0;
   ```